samples/nrf9160/modem_fault/src/main.c

```diff
@@ -0,0 +1,197 @@
+	"\r\n"
+
+#define HTTP_GET_LEN (sizeof(HTTP_GET) - 1)
+#define HTTP_HDR_END "\r\n\r\n"
+
+static const char send_buf[] = HTTP_GET;
+static char recv_buf[2048];
+
+/* Modem library initialization semaphore */
+static K_SEM_DEFINE(modem_sem, 0, 1);
+/* Network registration semaphore, given by main thread */
+static K_SEM_DEFINE(cereg_sem, 0, 1);
+
+/* Workqueue task crashing the modem by shutting it down
+ * while a download is in progress, without first de-registring from the network.
+ *
+ * In "normal" crash scenarions, the modem library remains initialized
+ * until it is shutdown by the default fault handler or by the application.
+ * As long as it remains initialized, it is possible to read out any
+ * pending network data that was delivered by modem before it crashed.
+ */
+static void crash_task(struct k_work *unused)
+{
+	(void) nrf_modem_shutdown();
+}
+
+static K_WORK_DELAYABLE_DEFINE(crash_work, crash_task);
+
+/* Modem library initialization hook */
+NRF_MODEM_LIB_ON_INIT(on_init, on_modem_init, NULL);
+
+static void on_modem_init(int err, void *ctx)
+{
+	if (err) {
+		LOG_INF("on_init: initialization failed, err %d", err);
+		return;
+	}
+
+	LOG_INF("on_init: modem library initialized");
+
+	/* Let main thread continue and register the device
+	 * to the LTE network upon modem re-initialization.
+	 */
+	k_sem_give(&modem_sem);
+}
+
+static void network_thread(void *p1, void *p2, void *p3)
+{
+	int err;
+	int fd;
+	size_t off;
+	ssize_t bytes;
+	struct addrinfo *res;
+	struct addrinfo hints = {
+		.ai_family = AF_INET,
+		.ai_socktype = SOCK_STREAM,
+		.ai_protocol = IPPROTO_TCP,
+	};
+
+	bool has_faulted = false;
+
+wait_for_cereg:
+	LOG_INF("net: waiting for network registration");
+	k_sem_take(&cereg_sem, K_FOREVER);
+
+	err = getaddrinfo(HTTP_HOSTNAME, HTTP_PORT, &hints, &res);
+	if (err) {
+		LOG_ERR("net: getaddrinfo() failed, err %d", errno);
+		return;
+	}
+
+	fd = socket(res->ai_family, res->ai_socktype, res->ai_protocol);
+	if (fd == -1) {
+		LOG_ERR("net: failed to open socket!");
+		return;
+	}
+
+	LOG_INF("net: connecting to %s", HTTP_HOSTNAME);
+	err = connect(fd, res->ai_addr, res->ai_addrlen);
+	if (err) {
+		LOG_ERR("net: connect() failed, err: %d", errno);
+		return;
+	}
+
+	(void)freeaddrinfo(res);
+
+	off = 0;
+	do {
+		bytes = send(fd, &send_buf[off], HTTP_GET_LEN - off, 0);
+		if (bytes < 0) {
+			LOG_ERR("net: send() failed, err %d", errno);
+			return;
+		}
+		off += bytes;
+	} while (off < HTTP_GET_LEN);
+
+	LOG_INF("net: sent %d bytes", off);
+
+	/* Schedule a crash during the download */
+	if (!has_faulted) {
+		has_faulted = true;
+		LOG_INF("net: crashing in two seconds");
+		k_work_schedule(&crash_work, K_SECONDS(2));
+	}
+
+	LOG_INF("net: downloading..");
+	off = 0;
+	do {
+		bytes = recv(fd, &recv_buf, sizeof(recv_buf), 0);
+		if (bytes < 0) {
+			LOG_WRN("net: recv() failed, err %d %s", errno, strerror(errno));
+			(void)close(fd);
+			/* Loop and wait for CEREG */
+			goto wait_for_cereg;
```

Comment:
I think not, for simplicity. Probably we want to state clearly that `EBADF` can only happen in this sample where we provoke the crash by shutting down the modem library (which closes all sockets and thus lets calls return with `EBADF`). Normally `EBADF` would not happen in case of modem crash.

Reply:
Yes, it would be good to add a comment that `EBADF` is only related to how modem restart is triggered here.

Just thinking whether this sample should utilize the shutdown callback. At least to me, that looks like the logical way to detect that modem has crashed. Relying on `errno` from `recv()` is probably not enough for real applications.